File: src/core/strategies/langgraph/repository.py
"""FhirRepository — typed data-access layer for FHIR queries.

Replaces ToolExecutor's string-keyed _dispatch() with real method calls.
Tools import this and call typed methods directly; no string routing.
"""

import json
import logging

from sqlalchemy.ext.asyncio import AsyncSession

from src.core.strategies.utils.constants import (
    DEFAULT_KEYWORD_LIMIT,
    DEFAULT_RESOURCE_LIMIT,
    MAX_SINGLE_RESULT_CHARS,
)
from src.core.strategies.utils.sql_guard import SQLValidationError, validate_sql
from src.db.queries import (
    execute_raw_sql,
    get_fhir_by_type,
    get_fhir_resources_schema_info,
    get_patient_overview,
    search_resources_by_keyword,
)

logger = logging.getLogger(__name__)

_TRUNCATION_MESSAGE = (
    "Result truncated. Use more specific filters to reduce result size."
)


def _truncate(data: dict | list, tool_name: str) -> str:
    """Serialize and truncate if over MAX_SINGLE_RESULT_CHARS."""
    json_str = json.dumps(data, default=str)
    if len(json_str) <= MAX_SINGLE_RESULT_CHARS:
        return json_str
    logger.warning(
        "tool_result_truncated | tool=%s | size=%d | cap=%d",
        tool_name,
        len(json_str),
        MAX_SINGLE_RESULT_CHARS,
    )
    return json.dumps(
        {
            "truncated": True,
            "message": _TRUNCATION_MESSAGE,
            "chars_returned": MAX_SINGLE_RESULT_CHARS,
            "total_chars": len(json_str),
        },
        default=str,
    )


class FhirRepository:
    """Typed FHIR data-access methods. One instance per tool invocation (holds db + patient_id)."""

    def __init__(self, db: AsyncSession, patient_id: str) -> None:
        self.db = db
        self.patient_id = patient_id

    async def patient_overview(self) -> tuple[str, list[str]]:
        """Returns (json_result, resource_types)."""
        try:
            data = await get_patient_overview(self.db, self.patient_id)
            types = [row["resource_type"] for row in data.get("by_type", [])]
            return _truncate(data, "overview"), types
        except Exception as e:
            logger.error("repo_error | method=patient_overview | error=%s", e)
            return json.dumps({"error": str(e)}), []

    async def resources_by_type(
        self,
        resource_type: str,
        limit: int = DEFAULT_RESOURCE_LIMIT,
    ) -> tuple[str, list[str], list[str]]:
        """Returns (json_result, resource_ids, resource_types)."""
        try:
            rows = await get_fhir_by_type(
                self.db, self.patient_id, resource_type, limit
            )
            ids = [r["resource_id"] for r in rows]
            types = [resource_type] if resource_type and rows else []
            return (
                _truncate({"resources": rows, "count": len(rows)}, "resources_by_type"),
                ids,
                types,
            )
        except Exception as e:
            logger.error(
                "repo_error | method=resources_by_type | type=%s | error=%s",
                resource_type,
                e,
            )
            return json.dumps({"error": str(e)}), [], []

    async def resources_by_keyword(
        self,
        keyword: str,
        limit: int = DEFAULT_KEYWORD_LIMIT,
    ) -> tuple[str, list[str], list[str]]:
        """Returns (json_result, resource_ids, resource_types)."""
        try:
            rows = await search_resources_by_keyword(
                self.db, self.patient_id, keyword, limit
            )
            ids = [r["resource_id"] for r in rows]
            types = list(
                {r["resource_type"] for r in rows if r.get("resource_type")}
            )
            return (
                _truncate({"resources": rows, "count": len(rows)}, "resources_by_keyword"),
                ids,
                types,
            )
        except Exception as e:
            logger.error(
                "repo_error | method=resources_by_keyword | keyword=%s | error=%s",
                keyword,
                e,
            )
            return json.dumps({"error": str(e)}), [], []

    async def resources_by_raw_sql(self, sql: str) -> tuple[str, list[str], list[str]]:
        """Returns (json_result, resource_ids, resource_types). Validates SQL first."""
        try:
            validated = validate_sql(sql)
        except SQLValidationError as e:
            return json.dumps({"error": str(e)}), [], []
        try:
            rows = await execute_raw_sql(
                self.db, validated, {"pid": self.patient_id}
            )
            ids = [
                str(r.get("resource_id") or r.get("id"))
                for r in rows
                if r.get("resource_id") or r.get("id")
            ]
            types = list(
                {r["resource_type"] for r in rows if r.get("resource_type")}
            )
            return (
                _truncate({"rows": rows, "count": len(rows)}, "resources_by_raw_sql"),
                ids,
                types,
            )
        except Exception as e:
            logger.error("repo_error | method=resources_by_raw_sql | error=%s", e)
            return json.dumps({"error": str(e)}), [], []

    async def fhir_resources_schema_info(self) -> str:
        """Returns JSON schema description of fhir_resources table."""
        try:
            data = await get_fhir_resources_schema_info(self.db)
            return json.dumps(data, default=str)
        except Exception as e:
            logger.error("repo_error | method=fhir_resources_schema_info | error=%s", e)
            return json.dumps({"error": str(e)})

    def finish_with_answer(self, answer: str, resource_ids: list[str]) -> str:
        """Package the final answer as a JSON string for ToolMessage content."""
        return json.dumps(
            {
                "answer": answer,
                "resource_ids": list(dict.fromkeys(resource_ids)),
                "resource_types": [],
            }
        )

File: ts-version/src/core/strategies/langgraph/repository.ts
// FhirRepository — typed data-access layer for FHIR queries.
//
// Replaces ToolExecutor's string-keyed _dispatch() with real method calls.
// Tools import this and call typed methods directly; no string routing.

import { logger } from '../../../config/logging';
import { DatabasePool } from '../../../db/session';
import {
  getAllFHIRByPatient,
  getPatientOverview,
  getFHIRByType,
  searchResourcesByKeyword,
  executeRawSQL,
  getFHIRResourcesSchemaInfo,
  getFHIRUISummaryDataByPatientId
} from '../../../db/queries';
import {
  DEFAULT_KEYWORD_LIMIT,
  DEFAULT_RESOURCE_LIMIT,
  MAX_SINGLE_TOOL_CHARS
} from '../utils/constants';
import { validateSQL, SQLValidationError } from '../utils/sql_guard';
import { RESOURCE_CATEGORIES } from './tools';
import { descopeInlineReferences } from '../utils/descope_ui_link';

const TRUNCATION_MESSAGE = 'Result truncated. Use more specific filters to reduce result size.';

function truncate(data: any, toolName: string): string {
  /**Serialize and truncate if over MAX_SINGLE_TOOL_CHARS.*/
  const jsonStr = JSON.stringify(data, (key, value) =>
    value instanceof Date ? value.toISOString() : value
  );

  if (jsonStr.length <= MAX_SINGLE_TOOL_CHARS) {
    return jsonStr;
  }

  logger.warn(
    `tool_result_truncated | tool=${toolName} | size=${jsonStr.length} | cap=${MAX_SINGLE_TOOL_CHARS}`
  );

  return JSON.stringify({
    truncated: true,
    message: TRUNCATION_MESSAGE,
    chars_returned: MAX_SINGLE_TOOL_CHARS,
    total_chars: jsonStr.length,
  }, (key, value) =>
    value instanceof Date ? value.toISOString() : value
  );
}

export class FhirRepository {
  /**Typed FHIR data-access methods. One instance per tool invocation (holds db + patient_id).*/

  constructor(
    private db: DatabasePool,
    private patient_id: string
  ) { }

  async getPatientOverview(): Promise<[string, string[]]> {
    /**Returns (json_result, resource_types).*/
    try {
      const data = await getPatientOverview(this.db, this.patient_id);
      const types = (data.by_type || []).map((row: any) => row.resource_type);
      return [truncate(data, 'get_patient_overview'), types];
    } catch (error) {
      logger.error(
        `repo_error | method=get_patient_overview | patient_id=${this.patient_id} | error=${String(error)}`
      );
      throw new Error(`Failed to get patient overview for patient ${this.patient_id}: ${String(error)}`);
    }
  }

  async getResourcesByType(
    resource_type: string,
    limit: number = DEFAULT_RESOURCE_LIMIT
  ): Promise<[string, string[], string[]]> {
    /**Returns (json_result, resource_ids, resource_types).*/
    try {
      const rows = await getFHIRByType(this.db, this.patient_id, resource_type, limit);
      const ids = rows.map((r: any) => r.resource_id);
      const types = resource_type && rows.length > 0 ? [resource_type] : [];

      return [
        truncate({ resources: rows, count: rows.length }, 'get_resources_by_type'),
        ids,
        types,
      ];
    } catch (error) {
      logger.error(
        `repo_error | method=get_resources_by_type | patient_id=${this.patient_id} | type=${resource_type} | error=${String(error)}`
      );
      return [JSON.stringify({ error: String(error) }), [], []];
    }
  }

  async getResourcesByKeyword(
    keyword: string,
    limit: number = DEFAULT_KEYWORD_LIMIT
  ): Promise<[string, string[], string[]]> {
    /**Returns (json_result, resource_ids, resource_types).*/
    try {
      const rows = await searchResourcesByKeyword(this.db, this.patient_id, keyword, limit);
      const ids = rows.map((r: any) => r.resource_id);
      const types = [...new Set(rows.map((r: any) => r.resource_type).filter(Boolean))];

      return [
        truncate({ resources: rows, count: rows.length }, 'get_resources_by_keyword'),
        ids,
        types,
      ];
    } catch (error) {
      logger.error(
        `repo_error | method=get_resources_by_keyword | patient_id=${this.patient_id} | keyword=${keyword} | error=${String(error)}`
      );
      return [JSON.stringify({ error: String(error) }), [], []];
    }
  }

  async getResourcesByRawSQL(sql: string): Promise<[string, string[], string[]]> {
    /**Returns (json_result, resource_ids, resource_types). Validates SQL first.*/
    try {
      sql = validateSQL(sql);
    } catch (error) {
      if (error instanceof SQLValidationError) {
        return [JSON.stringify({ error: error.message }), [], []];
      }
      throw error;
    }

    try {
      const rows = await executeRawSQL(this.db, sql, [this.patient_id]);
      const ids = rows
        .map((r: any) => String(r.resource_id || r.id))
        .filter(Boolean);
      const types = [...new Set(rows.map((r: any) => r.resource_type).filter(Boolean))];

      return [
        truncate({ rows, count: rows.length }, 'get_resources_by_raw_sql'),
        ids,
        types,
      ];
    } catch (error) {
      logger.error(
        `repo_error | method=get_resources_by_raw_sql | patient_id=${this.patient_id} | error=${String(error)}`
      );
      return [JSON.stringify({ error: String(error) }), [], []];
    }
  }

  async getFhirResourcesSchemaInfo(): Promise<string> {
    /**Returns JSON schema description of fhir_resources table.*/
    try {
      const data = await getFHIRResourcesSchemaInfo(this.db);
      return JSON.stringify(data, (key, value) =>
        value instanceof Date ? value.toISOString() : value
      );
    } catch (error) {
      logger.error(
        `repo_error | method=get_fhir_resources_schema_info | patient_id=${this.patient_id} | error=${String(error)}`
      );
      throw error;
    }
  }

  getFinalAnswer(answer: string, resource_ids: string[]): string {
    /**Package the final answer as a JSON string for ToolMessage content.*/
    answer = descopeInlineReferences(answer);
    return JSON.stringify({
      answer,
      resource_ids: [...new Set(resource_ids)], // Deduplicate
      resource_types: [],
    }, (key, value) =>
      value instanceof Date ? value.toISOString() : value
    );
  }

  // async getFHIRUISummaryDataByPatientId(): Promise<string> {
  //   try {
  //     const data = await getFHIRUISummaryDataByPatientId(this.db, this.patient_id);
  //     return JSON.stringify(data, (key, value) =>
  //       value instanceof Date ? value.toISOString() : value
  //     );
  //   } catch (error) {
  //     logger.error('repo_error | method=getFHIRUISummaryDataByPatientId | error=%s', error);
  //     throw error;
  //   }
  // }


  // --- Build context: full details for relevant categories, summaries for others ---
  async buildContextForLinks(relevantCategories: string[]) {
    try {
      const fhirData = await getFHIRUISummaryDataByPatientId(this.db, this.patient_id);
      const context: any = {
        patient: {
          name: fhirData.name,
          gender: fhirData.gender,
          birth_date: fhirData.birth_date,
          patient_id: fhirData.patient_id,
        },
      };

      for (const cat of RESOURCE_CATEGORIES) {
        const data = fhirData[cat] || [];
        if (relevantCategories.includes(cat)) {
          // Full data with details for relevant categories
          context[cat] = data.map((item: any) => ({
            name: item.name,
          }));
        } else {
          // Lightweight summaries for non-relevant categories
          context[cat + '_summary'] = data.map((item: any) => ({
            name: item.name,
            start: item.start,
            ...(item.status && { status: item.status }),
          }));
        }
      }

      return { context };
    } catch (error) {
      logger.error('repo_error | method=buildContextForLinks | error=%s', error);
      throw error;
    }
  }


}

//getFHIRUISummaryDataByPatientId
